### PR TITLE
RFC: Add cond() for sparse matrices

### PR DESCRIPTION
**Pull request by mfasi**
This PR would add the function `cond()` for sparse matrices, using the algorithm described in JuliaLang/LinearAlgebra.jl#104.

### CODE REVIEW

**Comment by andreasnoack** [comment]
Great. I'll review later today.

File: base/sparse/linalg.jl [hunk 1]
```diff
@@ -499,7 +499,196 @@ function norm(A::SparseMatrixCSC,p::Real=2)
     throw(ArgumentError("invalid p-norm p=$p. Valid: 1, Inf"))
 end
 
-# TODO
+# TODO rank
+
+# cond
+function cond(A::SparseMatrixCSC, p::Real=1)
```
Comment (tkelman) [inline base/sparse/linalg.jl]: `cond` defaulting to different norms for sparse vs dense is a little iffy.

And I think `normestinv` would look a little nicer without the leading and trailing lines of whitespace inside the method - the spacing within the code is fine, but right after `function` and right before `end` doesn't serve much purpose.

Comment (mfasi) [inline base/sparse/linalg.jl]: @tkelman I agree, this bugs me as well. Still, defaulting to a non implemented feature `p=2` would be rather annoying.  As an alternative, we could enforce the use of both arguments when `A` is sparse. 

You're right, those blank lines aren't that useful. I'll delete them next time I amend the commit.

File: base/sparse/linalg.jl [hunk 2]
```diff
@@ -499,7 +499,196 @@ function norm(A::SparseMatrixCSC,p::Real=2)
...
+    n = max(sizeA[1],sizeA[2])
+    F = factorize(A)
+    t = min(t,n)
+    if ~isreal(t) || length(t) > 1 || t <= 0
+        throw(ArgumentError("number of blocks must be a positive integer"))
+    end
+    if t > n
+        throw(ArgumentError("number of blocks must not be greater than $n"))
+    end
+    old_est = 0
```
Comment (andreasnoack) [inline base/sparse/linalg.jl]: Variable not used

File: base/sparse/linalg.jl [hunk 3]
```diff
@@ -499,7 +499,196 @@ function norm(A::SparseMatrixCSC,p::Real=2)
...
+function normestinv{T}(A::SparseMatrixCSC{T}, t::Integer = 2)
+
+    maxiter = 5
+
+    # Check the input
+    sizeA = size(A)
+    n = max(sizeA[1],sizeA[2])
+    F = factorize(A)
+    t = min(t,n)
+    if ~isreal(t) || length(t) > 1 || t <= 0
```
Comment (andreasnoack) [inline base/sparse/linalg.jl]: `t` is restricted to be an `Integer` so the `isreal` check shouldn't be necessary and neither should the `length` check.

File: base/sparse/linalg.jl [hunk 4]
```diff
@@ -499,7 +499,196 @@ function norm(A::SparseMatrixCSC,p::Real=2)
...
+    end
+end
+
+function normestinv{T}(A::SparseMatrixCSC{T}, t::Integer = 2)
+
+    maxiter = 5
+
+    # Check the input
+    sizeA = size(A)
+    n = max(sizeA[1],sizeA[2])
```
Comment (andreasnoack) [inline base/sparse/linalg.jl]: Maybe just a single line with `n = max(size(A)...)`

File: base/sparse/linalg.jl [hunk 5]
```diff
@@ -499,7 +499,196 @@ function norm(A::SparseMatrixCSC,p::Real=2)
...
+    old_est = 0
+    ind = Array(Int64, n)
+    ind_hist = Array(Int64, maxiter * t)
+    S = zeros(T, n, t)
+
+    # Generate the block matrix
+    h = zeros(real(T), n)
+    X = Array(T, n, t)
+    for i = 1:n
+        X[i, 1] = 1
```
Comment (andreasnoack) [inline base/sparse/linalg.jl]: `X[:,1] = 1` should be as efficient as the loop

File: base/sparse/linalg.jl [hunk 6]
```diff
@@ -499,7 +499,196 @@ function norm(A::SparseMatrixCSC,p::Real=2)
...
+                    break
+                end
+            end
+        end
+    end
+    scale!(X, 1./n)
+
+    iter = 0
+    est_old = 0
+    est = 0
```
Comment (andreasnoack) [inline base/sparse/linalg.jl]: This causes the return type to be a `Union{Int64,Float64}` and there are several other places where the type of the variable changes during the function. You should call 
```jl
code_warntype(SparseMatrix.normestinv, (SparseMatrixCSC{Float64,Int64}, Int))
```
and see how many of the red words you can get rid of. The return type from `factorize` cannot be made unique, but it shouldn't infect the rest of the code. If it does, we'll have to add some `::T`s, but don't do this initially.

File: base/sparse/linalg.jl (hunk 1 above)
Comment (andreasnoack) [inline base/sparse/linalg.jl]: What about using `2` as the default here. Eventually, we should support the 2 norm and until we do, people won't accidentally get different norm than they expect.

**Comment by mfasi** [comment]
@andreasnoack Than you, I changed what you suggested and amended the commit.

File: base/sparse/linalg.jl [hunk 7]
```diff
@@ -499,7 +499,189 @@ function norm(A::SparseMatrixCSC,p::Real=2)
...
+    end
+end
+
+function normestinv{T}(A::SparseMatrixCSC{T}, t::Integer = 2)
+    maxiter = 5
+    # Check the input
+    n = max(size(A)...)
+    F = factorize(A)
+    t = min(t,n)
+    if length(t) > 1 || t <= 0
```
Comment (andreasnoack) [inline base/sparse/linalg.jl]: `length` is not necessary here because `t` is restricted to be an `Integer`.

Comment (mfasi) [inline base/sparse/linalg.jl]: Sorry, I overlooked it. Fixed and amended.

**Comment by andreasnoack** [comment]
Greta. I've just confirmed that the return type is uniquely defined for `Float64` and `Complex{Float64}` input. Thank you for the contribution.

**Comment by mfasi** [comment]
Great. I think this closes  #6485.

**Pull request merged**